### PR TITLE
[Optimizer] Refactor legal gird analysis

### PR DESCRIPTION
**Pull request by odjuricicTT**
PR contains the following:
* Rename to LegalLayoutAnalysis
* Add generation of both TILE and ROW_MAJOR layouts
    * Keep ROW_MAJOR disabled by default and add flag to enable when needed
* Explicitly generate all layout related parameters (except for data format).
* Adjust overrides to support partial output layout instead of every param (e.g. only override buffer type (dram / l1)
* Clean up optimizer mlir tests

Left to do:
* Add cpp unit tests for analysis and overrides
* Add more partial override tests

### CODE REVIEW

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:3]: Can this be marked as rename instead of delete+add so that only new code can be highlighted?

File: test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir
Comment (nobradovictt) [inline test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir:1]: I don't like how this looks, should be in format similar to named parameters in case all options are not specified.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:3]: I think that this is a git / github internal decision and that there is not way to force it. Do you know a way?

File: test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir
Comment (odjuricicTT) [reply test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir:1]: I agree. Are you suggesting that we have multiple options on syntax for overrides or that we switch to always named?

Named might get a bit verbose and we might run out of meaningful separator characters.

E.g:
`override-output-layout=add_1=buffer_type=l1:tensor_memory_layout=block_sharded:memory_layout=row_major:data_type=f32`

Comment (nobradovictt) [reply test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir:1]: Look how named params work, you can omit them if you specify all params in correct order, if overriding only partially you need to specify param names. I agree verbosity is getting out of hand, maybe we can shorten names a bit. Lets think about it more.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:3]: This might work https://jtemporal.com/renaming-files-in-git-the-right-way/

File: test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir
Comment (nobradovictt) [reply test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir:1]: We can shorten param names:
override-output-layout=add_1=bt=l1:tml=block_sharded:ml=row_major:dt=f32 and avoid them if all are specified in order:
override-output-layout=add_1=l1:block_sharded:row_major:f32

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:261]: I do not like this approach. Can we at the beginning load all overrides and not even add layout not compatible with overrides instead of adding all and then removing them. At the very minimal create a util function instead of this lambda remove_if approach.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (svuckovicTT) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:3]: Didn't know this, ty @nobradovictt

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (svuckovicTT) [inline lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:15]: Shouldn't this be camelCase?

**Review by svuckovicTT — APPROVED** [review]
Dialect-side changes lgtm!

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:261]: I started with that approach and implemented a solution where i first create a search space for each layout param, then apply overrides and then try to itterate over the cros product and generate legal solutions. This was very hard to read because it ended up having a lot of nested if to check for non compatible cases across parameters. So i decided to go back to this which is cleaner an less error prone. I understand we are trading of performance for readability here, but it seems logical atm?

Will extract into a seperate funciton.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:114]: No, TILE / ROW_MAJOR is always overriden in this path and is controled with the next if: `if (override.memoryLayout == Layout::Tile)`.

This checks dataType which is allowed to not be present and is copied over from the existing tensor if not overriden.

File: test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir
Comment (odjuricicTT) [reply test/ttmlir/Dialect/TTNN/optimizer/partial_output_layout_override.mlir]: As discussed offline, change to support any number of layout params with deducing type of param by value.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (odjuricicTT) [reply lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:3]: Tried this, did work on single commit, but shown the same on the whole PR. Might be because there are a lot of changes in the file.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:143]: Lets not mix different styles of bool operators.

File: lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp:137]: Pls rename override to non C++ keyword, like overrideValue.

**Review by nobradovictt — APPROVED** [review]

**Comment by odjuricicTT** [comment]
@nsmithtt @tapspatel @vprajapati-tt Can someone take a look at the PR, your approval is required. I think that the only change not reviewed is this file https://github.com/tenstorrent/tt-mlir/pull/1363/files#diff-694f570cfe79f8e3130bde9637283269c8c33048a68ce3b3fd15be810b11d961

Will update codeowners for this in the future.

**Review by tapspatel — APPROVED** [review]

File: test/unittests/Optimizer/TestOptimizerOverrides.cpp
Comment (nsmithtt) [inline test/unittests/Optimizer/TestOptimizerOverrides.cpp:19]: @odjuricicTT, I think this might have broken macOS build: https://github.com/tenstorrent/tt-mlir/actions/runs/12269439074/job/34232997171

```
OptimizerTests: CommandLine Error: Option 'override-output-layout' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

Perhaps this fixture gets instantiated many times?

File: test/unittests/Optimizer/TestOptimizerOverrides.cpp
Comment (nsmithtt) [inline test/unittests/Optimizer/TestOptimizerOverrides.cpp:89]: @odjuricicTT, see comment above, maybe this one is the culprit:

```
[ RUN      ] OutputLayoutOverrideTest.ParseMultipleOps
OptimizerTests: CommandLine Error: Option 'override-output-layout' registered more than once!
```

File: test/unittests/Optimizer/TestOptimizerOverrides.cpp
Comment (odjuricicTT) [reply test/unittests/Optimizer/TestOptimizerOverrides.cpp:19]: Seems so. Thanks for pointing it out. Will take a look asap.

Comment (nsmithtt) [reply test/unittests/Optimizer/TestOptimizerOverrides.cpp:19]: I tried a potential fix here: https://github.com/tenstorrent/tt-mlir/pull/1589

But I'm unable to repro locally, it feels like a build config issue actually, I think the usage of `parser.parse` is correct.

Comment (odjuricicTT) [reply test/unittests/Optimizer/TestOptimizerOverrides.cpp:19]: It seems to be some strange behaviour where if two specific tests exist, one of them fails. Once one is deleted, all tests pass. Also unable to repro locally. Will delete one of the tests for now to unblock further work, and take a deeper look later.

Temporary fix: https://github.com/tenstorrent/tt-mlir/pull/1598

Issue for future: https://github.com/tenstorrent/tt-mlir/issues/1597

**Pull request merged**
